### PR TITLE
[Sampling] Add explicit bounded sampling-mask mode

### PR DESCRIPTION
**Pull request by nanjiangwill**
## Why

#27408 added SGLang's native sampling-mask response, but reconstructed support after sampling and used finite `top_k` as its safety bound. #33593, now merged in `main`, made top-p-only requests valid and exposed sampling masks through the OpenAI chat extension. The remaining contract must therefore be general: top-p/min-p support is data-dependent, and SGLang's threshold primitives retain every token tied at the cutoff, so even finite top-k does not bound realized support.

This PR makes the response describe the distribution that actually sampled the token, then bounds only its transport. This is a general serving contract, not an RL-only path.

## Contract

For the post-filter weights `w` used to sample token `y`, let `S = {i | w[i] > 0}` and `C = --sampling-mask-max-tokens`:

- sampling itself is unchanged;
- if `|S| <= C`, the returned token IDs are the complete support `S`;
- exact mode is the default, so `|S| > C` aborts before `y` is committed;
- `sampling_mask_mode="bounded"` opts into at most `C` IDs and reports `output_token_sampling_mask_truncated=true`;
- bounded mode keeps the highest-weight prefix and retains `y` by replacing its final entry when necessary;
- the selected-token logprob is always `log(w[y] / sum(w))` over the full realized support, never over the truncated return;
- missing or internally inconsistent producer output is an internal error, not truncation.

The cap changes metadata only. It never changes the sampling distribution.

## Design

- Capture the producer's representation. PyTorch reuses its filtered weights and token permutation; FlashInfer uses SGLang's existing top-k/top-p renormalization primitives and min-p predicate; deterministic logprob sampling and greedy sampling use their native representations.
- Keep internal results tensor-backed. `SamplingMaskOutput` is a dataclass following the existing result-holder `map_device_tensors`/async-D2H path; `_SamplingMaskCapture` is a sampler-private `NamedTuple`.
- Maintain opted-in device row indices in `SamplingBatchInfo`, including filter/merge, so mixed batches restrict dense work and transfer to requested rows.
- Emit factual `OK`, `TRUNCATED`, or `INVALID` status in the sampler, reach TP/CP consensus through the existing process groups, and apply exact/bounded policy in the scheduler.
- Propagate the per-token truncation bit through streaming, multi-tokenizer output, sessions, PP, and PD disaggregation. PD reuses its existing 64-byte padded metadata row; the first two slots carry returned length and truncation.
- Reuse the OpenAI chat sampling-mask path merged in #33593 and forward its `sampling_mask_mode` into the existing native request field.
- Abort exact overflow before committing the sampled token and release every request-owned KV slot, including overlap-allocated slots without a committed token ID.

## Internal API changes

- Private `Sampler._sample_from_probs` now returns sampled IDs plus optional producer capture data.
- `top_k_top_p_min_p_sampling_from_probs_torch` keeps its existing token-only behavior by default; a keyword-only opt-in returns already-computed filtered data for capture.
- Private `_attach_*` helpers became `_build_*` because they now return a tensor holder. The post-hoc `_compute_sampling_mask_from_probs` reconstruction was removed because it was not the producer's source of truth.
- Request fields are placed so existing positional constructors and the tokenized Rust wire layout do not shift.
- The PD field changed from `output_token_sampling_mask_len` to `output_token_sampling_mask_metadata` because the same padded row now carries length and truncation. No layout class, one-use global size constant, or additional RDMA buffer is introduced.
- Existing PP `get_logprob_*` helper names remain unchanged; only their logits-output payload is widened.
- Existing optional backend KV-release hooks are used at the release site; no new base-worker lifecycle interface is added.

No public sampling function is removed or renamed.

## Performance

- Feature-off requests add Python flag/tuple plumbing only: no new tensor operation, dense allocation, D2H copy, PP tensor, or TP/CP collective.
- Opted-in rows require dense `O(R * V)` support work because current sampling primitives produce dense weights. Packing transfers `4 * R * C + 12 * R` bytes. A fixed cap avoids a GPU-to-CPU synchronization to discover a data-dependent width; finite top-k is not a valid bound because cutoff ties are retained.
- On one H200 at `V=128256`, `C=4096`, measured opt-in operator overhead was approximately 0.28-0.72 ms across FlashInfer joint, min-p, and PyTorch paths.
- Dense temporary storage scaled with opted-in rows: approximately 1.7 MB at `R=1`, 54 MB at `R=32`, and 107 MB at `R=64`. Bounded D2H payload p50 was approximately 53-55 us from `R=1` through `R=64` (16 KB to 1.05 MB).
- Same-node TP2/4/8 status consensus measured approximately 20-42 us across 1-1024 rows.

## Validation

- 295 focused tests passed, 2 skipped, with 75 subtests on one H200 using `lmsysorg/sglang:nightly-dev-cu13-20260817-d91c3682` and the rebased PR source.
- Seven end-to-end H200 serving contracts passed on the rebased tree: finite support/logprob consistency, exact overflow abort, bounded truncation signal, top-p-only policy, OpenAI chat bounded-mode propagation, invalid-mode rejection, and deterministic sampling preservation.
- A 4,927-check H200 operator matrix passed against independent support and selected-logprob oracles, including uniform cutoff ties and min-p input isolation.
- One-node H200 TP2/4/8 collective probes passed.
- Repository-native `pre-commit run --all-files` passed on current `main`, including registered-test validation, lint-checker tests, Rust formatting, and workspace clippy. `git diff --check` passed.

The existing CUDA/AMD end-to-end sampling-mask test remains registered in its current CI suites.

---
### CI States

Latest PR Test (Base): :x: [Run #32076643369](https://github.com/sgl-project/sglang/actions/runs/32076643369)
Latest PR Test (Extra): :x: [Run #32076642814](https://github.com/sgl-project/sglang/actions/runs/32076642814)

### CODE REVIEW

File: python/sglang/srt/layers/sampler.py [hunk 1]
```diff
@@ -273,141 +311,207 @@ def _sample_from_probs(
...
         )
-        selected_logprobs = torch.log(
-            selected_raw_probs.float()
-            / support_mass.float().clamp_min(torch.finfo(torch.float32).tiny)
+        overflow = realized_lengths > self.sampling_mask_max_tokens
+        statuses = torch.where(
+            invalid,
+            SamplingMaskStatus.INVALID,
+            torch.where(
+                overflow & self.sampling_mask_strict,
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/sampler.py]: Defaulting overflow to OK silently breaks the existing mask/logprob contract: IDs are truncated, the logprob still uses full support, and clients get no truncation signal. Keep strict overflow as the default or expose truncation and make approximation opt-in.

**Review by JustinTong0323 — COMMENTED** [review]
Two blocking correctness issues remain.

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 2]
```diff
@@ -236,6 +241,9 @@ def process_batch_result_prefill(
             logprob_pt = 0
 
             for i, (req, next_token_id) in enumerate(zip(batch.reqs, next_token_ids)):
+                sampling_mask_finish_reason = self.get_sampling_mask_finish_reason(
```
Comment (JustinTong0323) [inline python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: PP may skip output communication for a non-final pure chunk and pass `logits_output=None`; this call then hits the helper's assertion before the middle-chunk branch. Defer the lookup to token-producing chunks or handle skipped output explicitly.

Comment (nanjiangwill) [reply python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: Addressed. Sampling-mask status is now checked only for active token-producing prefill chunks, so a skipped non-final PP chunk can keep `logits_output=None`. The helper remains strict for token-producing chunks. Added a regression test for the skipped-output path.

File: python/sglang/srt/layers/sampler.py (hunk 1 above)
Comment (nanjiangwill) [reply python/sglang/srt/layers/sampler.py]: Addressed. `return_sampling_mask=true` is now strict by default: if realized support exceeds `--sampling-mask-max-tokens`, the request returns HTTP 400 before committing the sampled token. Truncation requires explicit `sampling_mask_mode="bounded"` opt-in, and each returned token includes output_token_sampling_mask_truncated.

In bounded mode, the selected-token logprob intentionally remains computed over the complete post-filter distribution, since that is the distribution actually sampled from. The cap bounds only returned metadata and never changes or renormalizes the sampling distribution.

**Comment by nanjiangwill** [comment]
General sampling-mask support without finite `top_k` is deferred to #35765. #35205 restores finite-top-k admission in the meantime; this prototype and its review history remain available here.
